source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,70 @@
+
+If Engine loses connection with the host and cannot retrieve TPM data from the host before the VM is started elsewhere, contingent changes to the TPM data while the VM was running are lost.
+
+There is apparently no guarantee that new swtpm versions can still read its old data.  That should be watched in testing.
+
+## User Experience
+
+Ideally, a TPM device would be added to each VM automatically and would be automatically available to the guest OS.  However, this is a new feature and we cannot guarantee that everything is correctly supported in the underlying platform (for example, snapshots with memory are not working as expected).  For that reason, it's better to not add TPM devices to VMs by default for now and make a cluster
+option to enable them for the given cluster.
+
+TODO: Description of the UI actions once they are implemented.
+
+## Changes to Engine and Vdsm
+
+There are no special changes needed in order to add a TPM device as such, it's just another kind of device added to Engine.
+
+But a very different thing is ensuring TPM data persistence.  TPM data is managed using an external tool, swtpm.  Once the VM domain XML is deleted on the host, the TPM data is deleted as well by libvirt and it must be restored before the VM is run again, whether on this or another host (live host migrations are handled transparently and don't require any special actions on the oVirt side).
+
+In order to support TPM data persistence, we must store the data in the Engine database and transfer it to the host before the VM is started and update it in the Engine database after the VM gets down.  We need the following Vdsm API changes to achieve that:
+
+- Adding a new VM.create parameter providing the data.
+- Adding a new parameter to VM.destroy call response containing the updated data.
+
+There is some fragility with that approach.  If VM.destroy call fails for any reason after the libvirt domain is undefined (and the corresponding TPM data is deleted by libvirt), the updated data will be lost.  But we don't have many alternatives: While we could make a separate API call to retrieve TPM data or attach them to VM stats, that would mean retrieving live data from swtpm, which may not be consistent.  Only data retrieved after the VM stops can be considered stable.  And even with a separate API call, if Engine loses contact with the host and starts the VM elsewhere, updates to TPM data since the VM start will be lost.  We cannot guarantee, without introducing complicated implementations, up-to-date TPM data if the environment doesn't work as expected.  Considering TPM data typically serves forensic purposes rather than storing critical information, we may accept the fact that VM guest OS repair is needed in case the data got changed and the environment experienced unexpected trouble (i.e. VM.destroy call hasn't been successful).  It may still cause trouble for HA VMs.
```

Comment:
I'm not sure the approach with VM.destroy is reliable enough itself even when not considering other problems. Considering that TPM data doesn't change that often under normal circumstances, it could be complemented with stats reporting. A periodic job would watch for TPM data changes and if the data changed & remained the same in the next cycle then it would be reported as part of VM stats. I'd need more knowledge about how things work in Engine to implement that correctly but I think the basic idea could work. What do you think?